docs/src/QuickStart/running-qat.md

```diff
@@ -0,0 +1,414 @@
+  ret void
+}
+```
+
+Obviously, this code will not run on the second computer since that machine does
+it makes use of a `z` gate. While we could rewrite the frontend code, the beauti
+of an intermediate represenation is that such tasks can be performed at this
+level and hence, the user does not need to worry about hardware compatibility of
+the frontend code.
+
+QAT allows to address this issue by using a concept known as replacement
+linking. Replacement linking is linking process by which one function call is
+replaced by another if the and only if the other function is present in the IR.
+To facilitate this, we annotate the function `__quantum__qis__z__body` as
+follows:
+
+```llvm
+declare void @__quantum__qis__z__body(%Qubit*) #1
+
+attributes #1 = { "replaceWith"="soft_z" }
+```
+
+This states taht if the function `soft_z` is available, replace all calls to
+`__quantum__qis__z__body` with `soft_z`. Passing the `main.ll` through QAT does
+not result in any changes. However, adding a `lib.ll` with following code:
+
+```llvm
+define void @soft_z(%Qubit* %q) {
+entry:
+  call void @__quantum__qis__h__body(%Qubit* %q)
+  call void @__quantum__qis__x__body(%Qubit* %q)
+  call void @__quantum__qis__h__body(%Qubit* %q)
+  ret void
+}
+```
+
+will replace the calls `__quantum__qis__z__body` in `main.ll`. This is done by
+invoking QAT as follows:
+
+```sh
+qat --apply -S --target-def target.yaml --validate main.ll lib.ll
+```
+
+From the repository root, the code for this example can be found in
+`qat/demos/ReplaceLinking`. In this concrete demo, calling
+
+```sh
+qat --apply -S --target-def target.yaml --no-always-inline --validate main.ll lib.ll
+```
+
+produces
+
+```llvm
+; ModuleID = 'qat-link'
+source_filename = "qat-link"
+
+%Qubit = type opaque
+
+define void @example() #0 {
+entry:
+  call void @soft_z(%Qubit* null)
+  call void @__quantum__qis__cnot__body(%Qubit* null, %Qubit* nonnull inttoptr (i64 1 to %Qubit*))
+  ret void
+}
+
+declare %Qubit* @__quantum__rt__qubit_allocate()
+
+declare void @__quantum__qis__cnot__body(%Qubit*, %Qubit*)
+
+declare void @__quantum__rt__qubit_release(%Qubit*)
+
+define void @soft_z(%Qubit* %q) #1 {
+entry:
+  call void @__quantum__qis__h__body(%Qubit* %q)
+  call void @__quantum__qis__x__body(%Qubit* %q)
+  call void @__quantum__qis__h__body(%Qubit* %q)
+  ret void
+}
+
+declare void @__quantum__qis__h__body(%Qubit*)
+
+declare void @__quantum__qis__x__body(%Qubit*)
+
+attributes #0 = { "EntryPoint" "maxQubitIndex"="1" "maxResultIndex"="0" "requiredQubits"="2" "requiredResults"="0" }
+attributes #1 = { "maxQubitIndex"="0" "maxResultIndex"="0" "requiredQubits"="0" "requiredResults"="0" }
+```
+
+as would be expected. We note that we have disabled inlining as to maintain the
+function `soft_z`.
+
+Each backend can now provide a library that makes software implementations of
+the gates that are not supported natively by the system.
+
+### Circuit separation
```

Comment:
Circuit is such an overloaded term. I liked grouping better as a term. I commented in the code regarding some inconsistency in naming for this functioality.

Reply:
OK - we will still with grouping